lib/LemmaBucket.thy

```diff
@@ -482,4 +486,44 @@ lemma sorted_list_of_set_already_sorted:
   "\<lbrakk> distinct xs; sorted xs \<rbrakk> \<Longrightarrow> sorted_list_of_set (set xs) = xs"
   by (simp add: sorted_list_of_set_sort_remdups distinct_remdups_id sorted_sort_id)
 
+lemma inj_on_domD: "\<lbrakk>inj_on f (dom f); f x = Some z; f y = Some z\<rbrakk> \<Longrightarrow> x = y"
+  by (erule inj_onD) clarsimp+
+
+lemma not_emptyI:
+  "\<And>x A B. \<lbrakk>x\<in>A; x\<in>B\<rbrakk> \<Longrightarrow> A \<inter> B \<noteq> {}"
+  by auto
+
+lemma add_mask_lower_bits2:
+  "\<lbrakk>is_aligned (x :: 'a :: len word) n; p && ~~ mask n = 0\<rbrakk> \<Longrightarrow> x + p && ~~ mask n = x"
+  apply (subst word_plus_and_or_coroll)
+  apply (simp add: aligned_mask_disjoint and_mask_0_iff_le_mask)
+   apply (clarsimp simp: word_bool_alg.conj_disj_distrib2)
+  done
+
+(* This generalises pd_shifting' in ARM and ARM_HYP *)
+lemma pd_shifting_gen:
+  "\<lbrakk>b \<le> a; size pd - c \<le> a - b; is_aligned pd c \<rbrakk> \<Longrightarrow> pd + (vptr >> a << b) && ~~ mask c = pd"
+  apply (subgoal_tac "(vptr >> a << b) && ~~ mask c = 0")
+   apply (subst word_plus_and_or_coroll)
+    apply (erule aligned_mask_disjoint)
+    apply (simp add: and_mask_0_iff_le_mask[symmetric])
+   apply (simp add: word_bool_alg.conj_disj_distrib2)
+  apply (simp add: shiftr_shiftl1 neg_mask_twice word_bool_alg.conj.assoc)
+  apply (rule shiftr_not_mask_0)
+  apply (fastforce simp: max_def word_size)
+  done
```

Comment:
This one is a bit specific for lib, but there isn't an ARM*-common area. It might be worth considering which of these could go into Word_Lib, but maybe as a future next step.

Reply:
If it's too specific for lib, there's probably somewhere in ainvs that I could find a place for it. I'll try and do that.

And yes, not sure which of these can move up to Word_Lib or similar, but I'm glad you're happy to leave that for a future step.